File: 6.005/lec16/src/music/MusicLanguage.java
package music;

import static music.Pitch.OCTAVE;

import java.util.HashSet;
import java.util.Set;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * MusicLanguage defines static methods for
 * constructing and manipulating Music expressions, 
 * particularly to create
 * recursive music like rounds, canons, and fugues.
 */
public class MusicLanguage {
    // Prevent instantiation
    private MusicLanguage() {}
    
    ////////////////////////////////////////////////////
    // Factory methods
    ////////////////////////////////////////////////////
    
    /**
     * Make Music from a string using a variant of abc notation
     *  (see http://www.walshaw.plus.com/abc/examples/).
     * The notation consists of whitespace-delimited symbols representing either
     * notes or rests. The vertical bar | may be used as a delimiter 
     * for measures; make() treats it as a space.
     * Grammar:
     * notes ::= symbol*
     * symbol :: = . duration     for a rest
     *             pitch duration for a note
     *      pitch :: = accidental letter octave*
     *              accidental ::= empty string for natural,
     *                                    _ for flat,
     *                                    ^ for sharp
     *              letter ::= one of A-G                      
     *              octave ::= ' to raise one octave
     *                         , to lower one octave
     *      duration ::= empty string for one-beat duration,
     *                   /n for 1/n beat,
     *                   n for n-beat duration,
     *                   n/m for n/m-beat duration
     * Examples (assuming 4/4 common time, i.e. 4 beats per measure):
     *     C     quarter note, middle C
     *     A'2   half note, high A  
     *     _D/2  eighth note, middle D flat
     * 
     * @param notes string of notes and rests in simplified abc notation given above
     * @param instr instrument to play the notes with
     */
    public static Music notes(String notes, Instrument instr) {
        Music m = new Rest(0);
        for (String sym : notes.split("[\\s|]+")) {
            if (!sym.isEmpty()) {
                m = concat(m, parseSymbol(sym, instr));
            }
        }
        return m;
    }
    
    // Parse a symbol into a Note or a Rest.
    private static Music parseSymbol(String symbol, Instrument instr) {
        Matcher m = Pattern.compile("([^/0-9]*)([0-9]+)?(/[0-9]+)?").matcher(symbol);
        if (!m.matches()) throw new IllegalArgumentException("couldn't understand " + symbol);

        String pitchSymbol = m.group(1);

        double duration = 1.0;
        if (m.group(2) != null) duration *= Integer.valueOf(m.group(2));
        if (m.group(3) != null) duration /= Integer.valueOf(m.group(3).substring(1));

        if (pitchSymbol.equals(".")) return new Rest(duration);
        else return new Note(duration, parsePitch(pitchSymbol), instr);
    }
    
    // Parse a symbol into a Pitch.
    private static Pitch parsePitch(String symbol) {
        if (symbol.endsWith("'")) return parsePitch(symbol.substring(0, symbol.length()-1)).transpose(OCTAVE);
        else if (symbol.endsWith(",")) return parsePitch(symbol.substring(0, symbol.length()-1)).transpose(-OCTAVE);
        else if (symbol.startsWith("^")) return parsePitch(symbol.substring(1)).transpose(1);
        else if (symbol.startsWith("_")) return parsePitch(symbol.substring(1)).transpose(-1);
        else if (symbol.length() != 1) throw new IllegalArgumentException("can't understand " + symbol);
        else return new Pitch(symbol.charAt(0));
    }
    
    
    /**
     * @param duration length of rest, must be >= 0
     * @return rest of duration beats
     */
    public static Music rest(double duration) {
        return new Rest(duration);
    }

    
    ////////////////////////////////////////////////////
    // Producers
    ////////////////////////////////////////////////////

    /**
     * @param m1 first piece of music
     * @param m2 second piece of music
     * @return m1 followed by m2
     */
    public static Music concat(Music m1, Music m2) {
        return new Concat(m1, m2);
    }
        
    /**
     * @param m1 first piece of music
     * @param m2 second piece of music
     * @return m1 played at the same time as m2
     */
    public static Music together(Music m1, Music m2) {
        return new Together(m1, m2);
    }
    
    /**
     * @param m music to loop forever
     * @return music that repeatedly plays m in an endless loop
     */
    public static Music forever(Music m) {
        return new Forever(m);
    }
        

    
    ////////////////////////////////////////////////////
    // More operations on Music
    ////////////////////////////////////////////////////
    
    /**
     * param m piece of music
     * @returns set of instruments used by m
     */
    public static Set<Instrument> instrumentsUsed(Music m) {
        final Set<Instrument> instruments = new HashSet<Instrument>();        
        m.accept(new VoidVisitor () {
            public Void on(Note m) {
                instruments.add(m.instrument());
                return null;
            }
        });
        return instruments;
    }
    
    /**
     * Pause before playing some music.
     * @param m piece of music
     * @param beats length of delay in beats, must be >= 0
     * @returns rest of duration beats followed by m
     */
    public static Music delay(Music m, double beats) {
        return concat(new Rest(beats), m);
    }
    
    /**
     * Speed up or slow down a piece of music by a factor of speedup.
     * For example, changeTempo(m,2) returns music that plays twice as fast as m.
     * @param m piece of music
     * @param speedup factor to increase speed of music, must be > 0
     * @returns m' such that m'.duration() = m.duration()/speedup
     */
    public static Music changeTempo(Music m, final double speedup) {
        return m.accept(new IdentityVisitor() {
            @Override
            public Music on(Note m) {
                return new Note(m.duration()/speedup, m.pitch(), m.instrument());
            }
            @Override
            public Music on(Rest m) {
                return new Rest(m.duration()/speedup);
            }
        });
    }
    
    /**
     * Change all the notes in a piece of music to use a single instrument. 
     * For example, changeInstrument(m,PIANO) returns music that is played
     * entirely by the piano.
     * @requires m, instr != null
     * @returns m' such that for all notes n in m', n.instrument() == instr,
     *   but otherwise m' is identical to m 
     */
    public static Music changeInstrument(Music m, final Instrument instr) {
        return m.accept(new IdentityVisitor() {
            @Override
            public Music on(Note m) {
                return new Note(m.duration(), m.pitch(), instr);
            }
        });
    }
    
    /**
     * Replace all notes that use oldInstr with notes that use newInstr instead. 
     * @requires m, oldInstr, newInstr != null
     * @returns m' such that for all notes n in m such that n.instrument() == newInstr,
     *    the corresponding note n' in m' has n'.instrument() == newInstr.
     *   Otherwise m' is identical to m 
     */
    public static Music replaceInstrument(Music m, final Instrument oldInstr, final Instrument newInstr) {
        return m.accept(new IdentityVisitor() {
            @Override
            public Music on(Note m) {
                if (m.instrument().equals(oldInstr)) return new Note(m.duration(), m.pitch(), newInstr);
                else return m;
            }
        });
    }

    /**
     * Transpose all notes upward or downward in pitch.
     * @requires m != null
     * @returns m' such that for all notes n in m, the corresponding note n' in m'
     *  has n'.pitch() == n.pitch().transpose(semitonesUp).  Otherwise m' is identical
     *  to m.
     */
    public static Music transpose(Music m, final int semitonesUp) {
        return m.accept(new IdentityVisitor() {
            @Override
            public Music on(Note m) {
                return new Note(m.duration(), m.pitch().transpose(semitonesUp), m.instrument());
            }
        });
    }
    
}

File: 6.005/lec16/src/music/MusicLanguage.py
from music.MusicLanguage import *
from music.Instrument import *
from music import Pitch
from music.Pitch import OCTAVE
from java.lang.Double import POSITIVE_INFINITY # used for duration of forever()

#
#    ////////////////////////////////////////////////////////////////////
#    // Playing the music with MIDI 
#    ////////////////////////////////////////////////////////////////////
#

from music.midi import MusicPlayer

def play(music):
    ''' play music with the MIDI synthesizer '''
    MusicPlayer().play(music)


#
#    ////////////////////////////////////////////////////////////////////
#    // General higher-order functions 
#    ////////////////////////////////////////////////////////////////////
#

def identity(x):
    '''identity:X->X is the identity function.'''
    return x

def compose(f, g):
    '''f:A->B, g:B->C; returns h:A->C such that h=f o g.'''
    return lambda x: g(f(x))

def repeated(f, n):
    '''f:X->X, n:int >= 0.  Returns f^n, i.e. f composed with itself n times.'''
    if n==0:
        return identity;
    else:
        return compose(repeated(f, n-1), f)

def series(e, binop, f, n):
    '''e:E, binop:ExE->E, f:E->E, n:int >= 1; also requires binop to be associative.
       Returns e binop f(e) binop f^2(e) binop ... binop f^{n-1}(e).'''
    return reduce(binop, reduce(lambda fs,i: fs + [f(fs[-1])], range(1, n), [e]))
    
#
#    ////////////////////////////////////////////////////////////////////
#    // Functional objects that transform Music in interesting ways
#    ////////////////////////////////////////////////////////////////////
#    

def delayer(beats):
    '''beats:int >= 0; returns f:Music->Music that delays music by that number of beats'''
    return lambda music: delay(music, beats)

def tempoChanger(speedup):
    '''speedup:number > 0; returns f:Music->Music that speeds up music by a factor m'''
    return lambda music: changeTempo(music, speedup)

def instrumentChanger(instr):
    '''instr:Instrument; returns f:Music->Music that plays all the music with instr.'''
    return lambda music: changeInstrument(music, instr)

def instrumentReplacer(oldInstr, newInstr):
    '''oldInstr,newInstr:Instrument; returns f:Music->Music that replaces notes played by oldInstr
       with the same note played by newInstr.'''
    return lambda music: replaceInstrument(music, oldInstr, newInstr) 

def instrumentSequence(instrs):
    '''instrs:list(Instrument); returns f:Music->Music that replaces instrs[i] with instrs[i+1].'''
    def swapArgs(f): return lambda a,b: f(b, a)
    return reduce(swapArgs(compose), map(instrumentReplacer, instrs[:-1], instrs[1:]), identity)

def transposer(semitonesUp):
    '''semitonesUp:int; returns f:Music->Music that transposes music upward by semitonesUp.'''
    return lambda music: transpose(music, semitonesUp)

octave_higher = transposer(OCTAVE)

octave_lower = transposer(-OCTAVE)


#    //////////////////////////////////////////////////////////////////
#    // Operations for multiple voices: rounds, canons, counterpoint
#    //////////////////////////////////////////////////////////////////

def counterpoint(music, voices, f):
    '''music:Music, f:Music->Music, voice:int >= 1
       Returns n-voice contrapuntal composition 
       in which each voice i is given by f^i(m).'''
    return series(music, together, f, voices)

def canon(music, voices, beats, f=identity):
    '''music:Music, beats:int >= 0, f:Music->Music, voices:int >= 1
       Returns n-voice canon in which each voice i is given by f^i(m), 
       entering after i*beats.'''
    return counterpoint(music, voices, compose(f, delayer(beats)))


#
#    //////////////////////////////////////////////////////////////////
#    // Operations for repeating
#    //////////////////////////////////////////////////////////////////
#

def repeat(music, n, f=identity):
    '''music:Music, n:int >= 1, f:Music->Music.
       Returns n repetitions of music, where the ith repetition is f^{i-1}(music)'''
    return series(music, concat, f, n)

def accompany(music1, music2):
    '''music1,music2:Music.
       Requires music1 or music2 to run forever, or one's duration to be an even multiple
       of the other's duration.
       Returns a piece of music that plays music1 and music2 simultaneously,
       ending at the same time as well.'''
    if music1.duration() < music2.duration():
        return accompany(music2, music1)
    # so now music1.duration >= music2.duration
    elif music2.duration() == POSITIVE_INFINITY:
        return together(music1, music2)
    elif music1.duration() == POSITIVE_INFINITY:
        return together(music1, forever(music2))
    else:
        return together(music1, repeat(music2, round(m1.duration() / m2.duration())))
    

#    //////////////////////////////////////////////////////////////////
#    // Examples
#    //////////////////////////////////////////////////////////////////

# Row Row Row Your Boat
rowYourBoat = notes("""
   C C C3/4 D/4 E |
   E3/4 D/4 E3/4 F/4 G2 |
   C'/3 C'/3 C'/3 G/3 G/3 G/3 E/3 E/3 E/3 C/3 C/3 C/3 |
   G3/4 F/4 E3/4 D/4 C2
""", PIANO)        

# play it and then play it again, an octave higher
rowTwice = concat(rowYourBoat, transpose(rowYourBoat, OCTAVE))

# play it as a 4-voice round, each voice coming in after 4 beats
rowRound = canon(rowYourBoat, voices=4, beats=4)

# same 4-voice canon, but each voice an octave higher
rowOctaves = canon(rowYourBoat, voices=4, beats=4, f=transposer(OCTAVE))

# same thing, but repeated forever
rowForever = canon(forever(rowYourBoat), voices=4, beats=4, f=octave_higher);


# Frere Jacques
frereJacques = notes("""
    F G A F | F G A F | 
    A _B C'2 | A _B C'2 |
    C'/2 D'/2 C'/2 _B/2 A F |
    C'/2 D'/2 C'/2 _B/2 A F |
    F C F2 | F C F2
""", PIANO)
        
# 4-voice canon, come in after two measures, using four different instruments
frereRound = canon(frereJacques, voices=4, beats=frereJacques.duration()/4, f=instrumentSequence([PIANO, TRUMPET, ACCORDION, CHOIR_AAHS]))



# Pachelbel's canon
# The melody line below isn't complete -- for the rest, see http://www.musicaviva.com/ensemble/canon/music.tpl?filnavn=pachelbel-canon-3mndbc
pachelbelMelody = notes("""
    ^F'2 E'2 | D'2 ^C'2 | B2 A2 | B2 ^C'2 |
    D'2 ^C'2 | B2 A2 | G2 ^F2 | G2 E2 |
    D ^F A G | ^F D ^F E | D B, D A | G B A G |
    ^F D E ^C' | D' ^F' A' A | B G A ^F | D D' D3/2 .1/2 |
""", VIOLIN)
pachelbelCanon = canon(forever(pachelbelMelody), voices=3, beats=16)

# add a bass line, which starts by itself and then accompanies the melody
pachelbelBass = notes("D,2 A,,2 | B,,2 ^F,,2 | G,,2 D,,2 | G,,2 A,,2", CELLO);
pachelbel = concat(pachelbelBass, accompany(pachelbelCanon, pachelbelBass))


